hotxlfp/formulas/logic.py

```diff
@@ -7,69 +7,70 @@
 from . import error
 from . import utils
 import torch
+import numpy as np
 
 
-@dispatcher.register_for('AND')
+@dispatcher.register_for("AND")
 def AND(*args):
     args = utils.iflatten(args)
     return all(args)
 
 
-@dispatcher.register_for('IF')
+@dispatcher.register_for("IF")
 def IF(test, then, otherwise):
     if isinstance(test, error.XLError):
         return error.XLError
-    test_condition = torch.tensor(test, dtype=torch.bool)
-    if test_condition.all():
-        if isinstance(then, error.XLError):
-            return then
-        return torch.tensor(then, dtype=torch.double)
-    elif (~test_condition).all():
-        if isinstance(otherwise, error.XLError):
-            return otherwise
-        return torch.tensor(otherwise, dtype=torch.double)
-
     if isinstance(then, error.XLError):
         return then
     if isinstance(otherwise, error.XLError):
         return otherwise
-        
+
+    if (
+        isinstance(then, str)
+        or isinstance(otherwise, str)
+        or (isinstance(then, np.ndarray) and then.dtype.kind == "U")
+        or (isinstance(otherwise, np.ndarray) and otherwise.dtype.kind == "U")
+    ):
+        then_str = np.array(then, dtype="U")
+        otherwise_str = np.array(otherwise, dtype="U")
+        return np.where(np.array(test, dtype="b"), then_str, otherwise_str)
```

Comment:
This seems like a good strategy. I'll test to see if this np.array strategy works on our platform, but it also looks like it shouldn't cause any regressions because of the type checks you are doing.